### DISCUSSION TITLE
`INSERT` statement: How do I ignore superfluous columns from `SELECT` expression? (data import)

### DISCUSSION THREAD

**Discussion by derhuerst** [duckdb/duckdb#12410] [Q&A]
I want to import data from a (structurally valid) user-provided CSV file that has an unknown set of columns:
- It may not have all columns that I have defined in my table, but it seems this can be dealt with by adding a `BY NAME` clause to the `INSERT`.
- **It may have additional columns that are not present in the table. How do I ignore them**, and only pick those defined in the table?

From the docs:

> It’s important to note that when using `INSERT INTO ... BY NAME`, the column names specified in the `SELECT` statement must match the column names in the table. If a column name is misspelled or does not exist in the table, an error will occur. Columns that are missing from the `SELECT` statement will be filled with the default value.

## Example

Let's consider a `fruits.csv`:

```csv
name,weight
apple,0.2
banana,0.25
watermelon,1.3
```

```sql
CREATE TABLE fruits (
	name TEXT
	-- uncommenting makes the INSERT work
	-- weight REAL
);

INSERT INTO fruits
BY NAME
SELECT *
FROM read_csv('fruits.csv', header = true);
-- Binder Error: Table "fruits" does not have a column with name "weight"
```

---

```shell
duckdb -version
# v1.0.0 1f98600c2c
```

**Comment by Tishj** [comment]
I made an attempt, but it's hitting a Binder Exception
```sql
insert into tbl by name
select
	COLUMNS(
		x -> x IN (
			select
				list(column_name)
			from
				duckdb_columns()
			where table_name = 'tbl'
		)
	)
from (
	select
		'test',
		false,
		42,
		'error'
	)
t(a, b, c, d);
```

**Reply by Tishj** [reply]
doing a dynamic selection like this seems like it's not supported
You would have to use duckdb inside a general programming language (like Python) to construct your query, using something like the `duckdb_columns()` method I am using above to craft your selection list that only contains the column names that `fruits` has

**Reply by derhuerst** [reply]
True, this is an option. I'd love to be able to just generate DuckDB SQL and then – at any later point in time – execute it.

In addition, with the `duckdb_columns()`, I would have to read `fruits.csv` twice, at least partially, if I'm not mistaken. This is not a showstopper, but I'd like to avoid it.

**Reply by derhuerst** [reply]
I just discovered [`DESCRIBE TABLE 'fruits.csv'`](https://duckdb.org/docs/guides/meta/describe.html#describing-remote-tables), but it doesn't solve the problem either.

**Reply by Tishj** [reply]
I think using two DESCRIBE queries on `fruits.csv` and on`fruits`, finding the overlapping columns and then writing a query that projects only those columns is the most efficient way to do this

**Reply by derhuerst** [reply]
Do you know how to dynamically `SELECT` a set of columns?

**Reply by Tishj** [reply]
In what way?
Because that's kind of what I was trying to do with COLUMNS(...)

**Comment by Arengard** [comment]
INSERT OR IGNORE
INTO fruits
BY NAME
SELECT *
FROM read_csv('fruits.csv', header = true)

https://duckdb.org/docs/sql/statements/insert.html

**Reply by derhuerst** [reply]
Someone replied in [another comment](https://github.com/duckdb/duckdb/discussions/12410#discussioncomment-9811066):

> No, insert or ignore doesn't cover the case where the csv has a column which is not present in the target table.

**Comment by tonne1** [comment]
No, insert or ignore doesn't cover the case where the csv has a column which is not present in the target table.

**Reply by Arengard** [reply]
```
def insert_from_cvs_into_existing_table(con, table_name='csv', csv='test.csv'):
    # Execute the query to get the column names
    result = con.execute(f"select column_name from duckdb_columns() where table_name = '{table_name}'")
    column_names = [row[0] for row in result.fetchall()]

    # Construct the column selection part of the SQL query
    columns = ', '.join(column_names)

    # Construct the SQL query
    sql_query = f"SELECT {columns} FROM {csv}"

    # Execute the SQL query and fetch the result
    result = con.execute(sql_query)
    
    insrt_stmt = f"INSERT INTO {table_name} BY NAME SELECT {columns} from {csv}"
    try:
        con.begin()
        con.execute(insrt_stmt)
        con.commit()
    except Exception as e:
        print(f"An error occurred: {e}")
        con.rollback()
```

**Reply by derhuerst** [reply]
I'm generating a single blob of SQL, to be executed by DuckDB directly via stdin, so querying the set of columns and then generating the query is not an option for me.

**Reply by Tishj** [reply]
I think the conclusion is that that's not possible currently.
